File: src/main/java/no/nav/veilarbaktivitet/brukernotifikasjon/BrukernotifikasjonService.java
package no.nav.veilarbaktivitet.brukernotifikasjon;

import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import no.nav.veilarbaktivitet.aktivitet.AktivitetDAO;
import no.nav.veilarbaktivitet.arena.model.ArenaId;
import no.nav.veilarbaktivitet.manuell_status.v2.ManuellStatusV2Client;
import no.nav.veilarbaktivitet.manuell_status.v2.ManuellStatusV2DTO;
import no.nav.veilarbaktivitet.oppfolging.periode.SistePeriodeService;
import no.nav.veilarbaktivitet.person.Person;
import no.nav.veilarbaktivitet.person.PersonService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.scheduling.annotation.EnableScheduling;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.net.URL;
import java.util.Optional;
import java.util.UUID;

@Service
@Slf4j
@EnableScheduling
public class BrukernotifikasjonService {
    private final Logger secureLogs = LoggerFactory.getLogger("SecureLog");

    private final PersonService personService;
    private final SistePeriodeService sistePeriodeService;
    private final BrukerNotifikasjonDAO dao;

    private final ManuellStatusV2Client manuellStatusClient;
    private final String aktivitetsplanBasepath;
    private final AktivitetDAO aktivitetDAO;

    public BrukernotifikasjonService(PersonService personService, SistePeriodeService sistePeriodeService, BrukerNotifikasjonDAO dao, ManuellStatusV2Client manuellStatusClient, @Value("${app.env.aktivitetsplan.basepath}") String aktivitetsplanBasepath,AktivitetDAO aktivitetDAO) {
        this.personService = personService;
        this.sistePeriodeService = sistePeriodeService;
        this.dao = dao;
        this.manuellStatusClient = manuellStatusClient;
        this.aktivitetsplanBasepath = aktivitetsplanBasepath;
        this.aktivitetDAO = aktivitetDAO;
    }


    public void setDone(
            long aktivitetId,
            VarselType varseltype
    ) {
        dao.setDone(aktivitetId, varseltype);
    }

    public void setDone(
            ArenaId aktivitetId,
            VarselType varseltype
    ) {
        dao.setDone(aktivitetId, varseltype);
    }

    public boolean kanVarsles(
            Person.AktorId aktorId
    ) {
        Optional<ManuellStatusV2DTO> manuellStatusResponse = manuellStatusClient.get(aktorId);

        boolean erManuell = manuellStatusResponse.map(ManuellStatusV2DTO::isErUnderManuellOppfolging).orElse(true);
        boolean erReservertIKrr = manuellStatusResponse.map(ManuellStatusV2DTO::getKrrStatus).map(ManuellStatusV2DTO.KrrStatus::isErReservert).orElse(true);


        boolean kanVarsles = !erManuell && !erReservertIKrr;
        if(!kanVarsles) {
            secureLogs.info("bruker kan ikke varsles aktorId: {}, erManuell: {}, erReservertIKrr: {}", aktorId.get(), erManuell, erReservertIKrr);
        }

        return kanVarsles;
    }

    @Transactional
    public UUID opprettVarselPaaAktivitet(
            long aktivitetId,
            long aktitetVersion,
            Person.AktorId aktorId,
            String ditNavTekst,
            VarselType varseltype
    ) {
        return opprettVarselPaaAktivitet(
                aktivitetId,
                aktitetVersion,
                aktorId,
                ditNavTekst,
                varseltype,
                null, null, null //Disse settes til standartekst av brukernotifiaksjoenr hvis ikke satt
        );
    }

    @Transactional
    public UUID opprettVarselPaaAktivitet(
            long aktivitetId,
            long aktitetVersion,
            Person.AktorId aktorId,
            String ditNavTekst,
            VarselType varseltype,
            String epostTitel,
            String epostBody,
            String smsTekst
    ) {
        UUID uuid = UUID.randomUUID();

        Person.Fnr fnr = personService
                .getFnrForAktorId(aktorId);

        UUID gjeldendeOppfolgingsperiode = sistePeriodeService.hentGjeldendeOppfolgingsperiodeMedFallback(aktorId);
        URL aktivtetUrl = createAktivitetLink(aktivitetId + "");

        long brukernotifikasjonId = dao.opprettBrukernotifikasjon(uuid, fnr, ditNavTekst, gjeldendeOppfolgingsperiode, varseltype, VarselStatus.PENDING, aktivtetUrl, epostTitel, epostBody, smsTekst);
        dao.kobleAktivitetIdTilBrukernotifikasjon(brukernotifikasjonId, aktivitetId, aktitetVersion);

        return uuid;

    }

    public boolean finnesBrukernotifikasjonMedVarselTypeForAktivitet(long aktivitetsId, VarselType varselType) {
        return dao.finnesBrukernotifikasjonMedVarselTypeForAktivitet(aktivitetsId, varselType);
    }

    @Transactional
    public UUID opprettVarselPaaArenaAktivitet(
            ArenaId arenaAktivitetId,
            Optional<Long> aktivitetId,
            Person.Fnr fnr,
            String ditNavTekst,
            VarselType varseltype
    ) {
        UUID uuid = UUID.randomUUID();

        Person.AktorId aktorId = personService
                .getAktorIdForPersonBruker(fnr)
                .orElseThrow();

        UUID gjeldendeOppfolgingsperiode = sistePeriodeService.hentGjeldendeOppfolgingsperiodeMedFallback(aktorId);
        URL aktivtetUrl = createAktivitetLink(aktivitetId.map(Object::toString).orElseGet(arenaAktivitetId::id));

        // epostTittel, epostBody og smsTekst settes til standartekst av brukernotifiaksjoenr hvis ikke satt
        long brukernotifikasjonId = dao.opprettBrukernotifikasjon(uuid, fnr, ditNavTekst, gjeldendeOppfolgingsperiode, varseltype, VarselStatus.PENDING, aktivtetUrl, null, null, null);
        dao.kobleArenaAktivitetIdTilBrukernotifikasjon(brukernotifikasjonId, arenaAktivitetId);
        // Populer brukernotifikasjon koblingstabell til vanlig aktivitet også
        aktivitetId
            .flatMap(aktivitetDAO::hentMaybeAktivitet)
            .ifPresent(aktivitet -> dao.kobleAktivitetIdTilBrukernotifikasjon(brukernotifikasjonId, aktivitet.getId(), aktivitet.getVersjon()));

        return uuid;

    }

    @SneakyThrows
    private URL createAktivitetLink(String aktivitetId) {
        return new URL(aktivitetsplanBasepath + "/aktivitet/vis/" + aktivitetId);
    }

    public void setDoneGrupperingsID(UUID uuid) {
        dao.setDoneGrupperingsID(uuid);
    }
}

File: src/main/kotlin/no/nav/veilarbaktivitet/brukernotifikasjon/BrukernotifikasjonService.kt
package no.nav.veilarbaktivitet.brukernotifikasjon

import io.micrometer.core.annotation.Timed
import no.nav.veilarbaktivitet.aktivitet.AktivitetDAO
import no.nav.veilarbaktivitet.arena.model.ArenaId
import no.nav.veilarbaktivitet.brukernotifikasjon.opprettVarsel.AktivitetVarsel
import no.nav.veilarbaktivitet.brukernotifikasjon.opprettVarsel.ArenaAktivitetVarsel
import no.nav.veilarbaktivitet.brukernotifikasjon.varsel.SkalSendes
import no.nav.veilarbaktivitet.brukernotifikasjon.varsel.VarselDAO
import no.nav.veilarbaktivitet.manuell_status.v2.ManuellStatusV2Client
import no.nav.veilarbaktivitet.oppfolging.periode.SistePeriodeService
import no.nav.veilarbaktivitet.person.Person
import no.nav.veilarbaktivitet.person.PersonService
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
open class BrukernotifikasjonService(
    val brukernotifikasjonProducer: BrukernotifikasjonProducer,
    val varselDAO: VarselDAO,
    val brukerNotifikasjonDAO: BrukerNotifikasjonDAO,
    val aktivitetDao: AktivitetDAO,
    val manuellStatusClient: ManuellStatusV2Client,
    val personService: PersonService,
    @Value("\${app.env.aktivitetsplan.basepath}")
    val aktivitetsplanBasepath: String,
    var sistePeriodeService: SistePeriodeService
) {
    private val log = LoggerFactory.getLogger(BrukernotifikasjonService::class.java)
    private val secureLogs: Logger = LoggerFactory.getLogger("SecureLog")

    open fun hentVarselSomSkalSendes(maxAntall: Int): List<SkalSendes> = varselDAO.hentVarselSomSkalSendes(maxAntall)
    open fun avbrytIkkeSendteOppgaverForAvslutteteAktiviteter(): Int = varselDAO.avbrytIkkeSendteOppgaverForAvslutteteAktiviteter()

    open fun finnesBrukernotifikasjonMedVarselTypeForAktivitet(aktivitetsId: Long, varselType: VarselType): Boolean {
        return brukerNotifikasjonDAO.finnesBrukernotifikasjonMedVarselTypeForAktivitet(aktivitetsId, varselType)
    }
    open fun setDone(aktivitetId: Long, varseltype: VarselType) = brukerNotifikasjonDAO.setDone(aktivitetId, varseltype)
    open fun setDone(aktivitetId: ArenaId, varseltype: VarselType) = brukerNotifikasjonDAO.setDone(aktivitetId, varseltype)
    open fun setDoneGrupperingsID(uuid: UUID) = brukerNotifikasjonDAO.setDoneGrupperingsID(uuid)

    @Transactional
    @Timed(value = "brukernotifikasjon_opprett_oppgave_sendt")
    open fun send(skalSendes: SkalSendes) {
        val oppdatertOk = varselDAO.setSendt(skalSendes.brukernotifikasjonLopeNummer)
        if (oppdatertOk) {
            sendVarsel(skalSendes)
        }
    }

    private fun sendVarsel(skalSendes: SkalSendes) {
        val offset: Long = brukernotifikasjonProducer.send(skalSendes)
        log.debug(
            "Brukernotifikasjon {} med type {} publisert med offset {}",
            skalSendes.brukernotifikasjonId.toString(),
            skalSendes.varselType.brukernotifikasjonType.name,
            offset
        )
    }

    open fun kanVarsles(aktorId: Person.AktorId): Boolean {
        val manuellStatusResponse = manuellStatusClient.get(aktorId)
        val erManuell = manuellStatusResponse.map { it.isErUnderManuellOppfolging }.orElse(true)
        val erReservertIKrr = manuellStatusResponse.map { it.krrStatus }.map { it.isErReservert }.orElse(true)

        val kanVarsles = !erManuell && !erReservertIKrr
        if (!kanVarsles) {
            secureLogs.info(
                "bruker kan ikke varsles aktorId: {}, erManuell: {}, erReservertIKrr: {}",
                aktorId.get(),
                erManuell,
                erReservertIKrr
            )
        }
        return kanVarsles
    }

    @Transactional
    open fun opprettVarselPaaAktivitet(varsel: AktivitetVarsel): UUID {
        val uuid = UUID.randomUUID()
        val fnr = personService.getFnrForAktorId(varsel.aktorId)
        val gjeldendeOppfolgingsperiode = sistePeriodeService.hentGjeldendeOppfolgingsperiodeMedFallback(varsel.aktorId)
        val brukernotifikasjonId = brukerNotifikasjonDAO.opprettBrukernotifikasjonIOutbox(
            varsel.toUgåendeVarsel(
                uuid,
                gjeldendeOppfolgingsperiode,
                aktivitetsplanBasepath,
                fnr
            )
        )
        brukerNotifikasjonDAO.kobleAktivitetIdTilBrukernotifikasjon(brukernotifikasjonId, varsel.aktivitetId, varsel.aktitetVersion)
        return uuid
    }

    @Transactional
    open fun opprettVarselPaaArenaAktivitet(
        arenaAktivitetVarsel: ArenaAktivitetVarsel
    ): UUID {
        val uuid = UUID.randomUUID()
        val aktorId = personService
            .getAktorIdForPersonBruker(arenaAktivitetVarsel.fnr)
            .orElseThrow()
        val gjeldendeOppfolgingsperiode = sistePeriodeService.hentGjeldendeOppfolgingsperiodeMedFallback(aktorId)

        // epostTittel, epostBody og smsTekst settes til standartekst av brukernotifiaksjoenr hvis ikke satt
        val brukernotifikasjonId = brukerNotifikasjonDAO.opprettBrukernotifikasjonIOutbox(
            arenaAktivitetVarsel.toUgåendeVarsel(
                uuid,
                gjeldendeOppfolgingsperiode,
                aktivitetsplanBasepath,
            )
        )
        brukerNotifikasjonDAO.kobleArenaAktivitetIdTilBrukernotifikasjon(brukernotifikasjonId, arenaAktivitetVarsel.arenaAktivitetId)
        // Populer brukernotifikasjon koblingstabell til vanlig aktivitet også
        arenaAktivitetVarsel.aktivitetId
            .flatMap { id -> aktivitetDao.hentMaybeAktivitet(id) }
            .ifPresent { aktivitet ->
                brukerNotifikasjonDAO.kobleAktivitetIdTilBrukernotifikasjon(
                    brukernotifikasjonId,
                    aktivitet.id,
                    aktivitet.versjon
                )
            }

        return uuid
    }
}
